pkg/traceutil/trace.go

```diff
@@ -0,0 +1,143 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package traceutil
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// Field is a kv pair to record additional details of the trace.
+type Field struct {
+	Key   string
+	Value interface{}
+}
+
+func (f *Field) format() string {
+	return fmt.Sprintf("%s:%v; ", f.Key, f.Value)
+}
+
+func writeFields(fields []Field) string {
+	if len(fields) == 0 {
+		return ""
+	}
+	var buf bytes.Buffer
+	buf.WriteString("{")
+	for _, f := range fields {
+		buf.WriteString(f.format())
+	}
+	buf.WriteString("}")
+	return buf.String()
+}
+
+type Trace struct {
+	operation string
+	fields    []Field
+	startTime time.Time
+	steps     []step
+}
+
+type step struct {
+	time   time.Time
+	msg    string
+	fields []Field
+}
+
+func New(op string, fields ...Field) *Trace {
+	return &Trace{operation: op, startTime: time.Now(), fields: fields}
+}
+
+// TODO returns a non-nil, empty Trace
+func TODO() *Trace {
+	return &Trace{}
+}
+
+func Get(ctx context.Context) *Trace {
+	if trace, ok := ctx.Value("trace").(*Trace); ok && trace != nil {
+		return trace
+	}
+	return TODO()
+}
+
+func GetOrCreate(ctx context.Context, op string, fields ...Field) (context.Context, *Trace) {
+	trace, ok := ctx.Value("trace").(*Trace)
+	if !ok || trace == nil {
+		trace = New(op)
+		trace.fields = fields
+		ctx = context.WithValue(ctx, "trace", trace)
+	}
+	return ctx, trace
+}
+
+func (t *Trace) Step(msg string, fields ...Field) {
+	t.steps = append(t.steps, step{time: time.Now(), msg: msg, fields: fields})
+}
+
+func (t *Trace) AddField(fields ...Field) {
+	for _, f := range fields {
+		t.fields = append(t.fields, f)
+	}
+}
+
+// Log dumps all steps in the Trace
+func (t *Trace) Log(lg *zap.Logger) {
+	t.LogWithStepThreshold(0, lg)
+}
+
+// LogIfLong dumps logs if the duration is longer than threshold
+func (t *Trace) LogIfLong(threshold time.Duration, lg *zap.Logger) {
+	if time.Since(t.startTime) > threshold {
+		stepThreshold := threshold / time.Duration(len(t.steps)+1)
+		t.LogWithStepThreshold(stepThreshold, lg)
+	}
+}
+
+// LogWithStepThreshold only dumps step whose duration is longer than step threshold
+func (t *Trace) LogWithStepThreshold(threshold time.Duration, lg *zap.Logger) {
+	msg, fs := t.logInfo(threshold)
+	if lg != nil {
+		lg.Info(msg, fs...)
+	}
+}
+
+func (t *Trace) logInfo(threshold time.Duration) (string, []zap.Field) {
+	endTime := time.Now()
+	totalDuration := endTime.Sub(t.startTime)
+	traceNum := rand.Int31()
+	msg := fmt.Sprintf("Trace[%d] %s", traceNum, t.operation)
```

Comment:
How about `trace[id=%d]`? Trying to use lower case whenever possible; think this is more readable when encoded in structure logging. what do you think?

Reply:
Sounds good to me. I updated the example output above.